language-support/ts/daml-ledger/index.ts

```diff
@@ -608,44 +628,147 @@ class Ledger {
+    function off(_t: any, _l: any): any {}
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    return { on, off, close: (): void => {} };
+  }
+
+  /**
+   * @internal
+   *
+   */
+  streamFetchByKeyCommon<T extends object, K, I extends string, S>(
+    template: Template<T, K, I>,
+    keys: K[], // non-zero length
+    name: string,
+    to: (contracts: (CreateEvent<T, K, I> | null)[]) => S,
+    from: (state: S) => (CreateEvent<T, K, I> | null)[],
+  ): Stream<T, K, I, S> {
+    const lastContractIds: (ContractId<T> | null)[] = Array(keys.length).fill(null);
+    const keysCopy = Array.from(keys);
+    const initState: (CreateEvent<T, K, I> | null)[] = Array(keys.length).fill(null);
+    const request = keys.map(k => ({templateId: template.templateId, key: k}));
+    const reconnectRequest = (): object[] => request.map((r, idx) => ({...r, 'contractIdAtOffset': lastContractIds[idx]}))
+    const change = (state: S, events: readonly Event<T, K, I>[]): S => {
+      const contracts = from(state);
+      const newState: (CreateEvent<T, K, I> | null)[] = Array.from(contracts);
       for (const event of events) {
         if ('created' in event) {
-          contract = event.created;
+          const k = event.created.key;
+          keysCopy.forEach((requestKey, idx) => {
+            if (deepEqual(requestKey, k, {strict: true})) {
+              newState[idx] = event.created;
+            }
+          });
         } else { // i.e. 'archived' in event
-          if (contract && contract.contractId === event.archived.contractId) {
-            contract = null;
-          }
+          const id: ContractId<T> = event.archived.contractId;
+          newState.forEach((contract, idx) => {
+            if (contract && contract.contractId === id) {
+              newState[idx] = null;
+            }
+          });
         }
       }
-      lastContractId = contract ? contract.contractId : null
-      return contract;
+      newState.forEach((c, idx) => {
+        lastContractIds[idx] = c ? c.contractId : null;
+      });
+      return to(newState);
+    }
+    return this.streamSubmit(name, template, 'v1/stream/fetch', request, reconnectRequest, to(initState), change);
+  }
+
+  /**
+   * Retrieve a consolidated stream of events for a list of keys and a single
+   * template.
+   *
+   * The accumulated state is an array of the same length as the given list of
+   * keys, with positional correspondence. Each element in the array represents
+   * the current contract for the given key, or is explicitly null if there is
+   * currently no active contract matching that key.
+   *
+   * @typeparam T The contract template type.
+   * @typeparam K The contract key type.
+   * @typeparam I The contract id type.
+   */
+  streamFetchByKeys<T extends object, K, I extends string>(
+    template: Template<T, K, I>,
+    keys: K[],
+  ): Stream<T, K, I, (CreateEvent<T, K, I> | null)[]> {
+    if (keys.length == 0) {
```

Comment:
Side note for myself: I briefly wondered if we should require >= 1 keys. However, that makes things harder if the list of keys is dynamic so I prefer the implementation here.

Reply:
That was also my reasoning for adding support for 0 keys. Maybe I should add a comment to that effect.